src/client/src/Stream.c

```diff
@@ -1836,6 +1843,153 @@ STATUS setNalAdaptationFlags(PKinesisVideoStream pKinesisVideoStream, UINT32 nal
     return retStatus;
 }
 
+/**
+ * Puts a metadata into the stream
+ */
+STATUS putEventMetadata(PKinesisVideoStream pKinesisVideoStream, UINT32 event, PStreamEventMetadata pMetadata)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PKinesisVideoClient pKinesisVideoClient = NULL;
+    BOOL streamLocked = FALSE, hasMetadata = pMetadata == NULL ? FALSE : TRUE, firstEvent = TRUE, creatingNodes = FALSE, tagsNeeded = TRUE;
+    UINT8 iter = 0;
+    UINT32 packagedSize = 0, totalPackagedSize = 0, metadataQueueSize;
+    UINT32 packagedSizes[MAX_FRAGMENT_METADATA_COUNT] = {0};
+    PSerializedMetadata serializedNodes[MAX_FRAGMENT_METADATA_COUNT] = {0};
+    UINT8 neededNodes = 0;
+    PSerializedMetadata pExistingSerializedMetadata;
+    StackQueueIterator iterator;
+    UINT64 data;
+
+    CHK(pKinesisVideoStream != NULL, STATUS_NULL_ARG);
+    pKinesisVideoClient = pKinesisVideoStream->pKinesisVideoClient;
+
+    // Check if the stream has been stopped
+    CHK(!pKinesisVideoStream->streamStopped, STATUS_STREAM_HAS_BEEN_STOPPED);
+
+    // Lock the stream
+    pKinesisVideoClient->clientCallbacks.lockMutexFn(pKinesisVideoClient->clientCallbacks.customData, pKinesisVideoStream->base.lock);
+    streamLocked = TRUE;
+
+    if (pKinesisVideoStream->metadataTracker.activeMkvEvent) {
+        // only the first event may have a L1 'tags' element.
+        firstEvent = FALSE;
+    }
+
+    // Check if we are in the right state only if we are not in a rotation state
+    if (pKinesisVideoStream->streamState != STREAM_STATE_READY) {
+        CHK_STATUS(acceptStateMachineState(pKinesisVideoStream->base.pStateMachine,
+                                           STREAM_STATE_READY | STREAM_STATE_PUT_STREAM | STREAM_STATE_TAG_STREAM | STREAM_STATE_STREAMING |
+                                               STREAM_STATE_GET_ENDPOINT | STREAM_STATE_GET_TOKEN | STREAM_STATE_STOPPED));
+    }
+
+    // Validate if the customer is not attempting to add an internal metadata
+    if (hasMetadata) {
+        for (iter = 0; iter < pMetadata->numberOfPairs; iter++) {
+            CHK(0 != STRNCMP(AWS_INTERNAL_METADATA_PREFIX, pMetadata->names[iter], (SIZEOF(AWS_INTERNAL_METADATA_PREFIX) - 1) / SIZEOF(CHAR)),
+                STATUS_INVALID_METADATA_NAME);
+        }
+        if (pMetadata->imagePrefix != NULL) {
+            CHK(0 != STRNCMP(AWS_INTERNAL_METADATA_PREFIX, pMetadata->imagePrefix, (SIZEOF(AWS_INTERNAL_METADATA_PREFIX) - 1) / SIZEOF(CHAR)),
+                STATUS_INVALID_METADATA_NAME);
+        }
+        neededNodes += pMetadata->numberOfPairs;
+    }
+
+    // Ensure we don't have more than MAX size of the metadata queue
+    if (event & STREAM_EVENT_TYPE_NOTIFICATION) {
```

Comment:
Can you create macros for this like we have for frame flags, see:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/mkvgen/include/com/amazonaws/kinesis/video/mkvgen/Include.h#L230-L233